torch/sparse/__init__.py

```diff
@@ -525,3 +525,178 @@ def func_wrapper(*args, **kwargs):
         return gradcheck(*args, **kwargs)
 
     return gradcheck_with_sparse_outputs
+
+
+def enable_sparse_support(gradcheck):
+    """Decorator for torch.autograd.gradcheck or its functools.partial
+    variants that extends the gradcheck function with support to input
+    functions that operate on or/and return sparse tensors. The specified
+    gradcheck function itself is guaranteed to operate on strided
+    tensors only."""
+
+    def gradcheck_with_sparse_support(func, inputs, **kwargs):
+        """Same as :func:`torch.autograd.gradcheck` but with sparse tensors
+        inputs and outputs support.
+
+        .. note::
+
+            When ``masked == False``, the gradients of unspecified
+            elements of sparse tensors are not ignored. This means
+            that the input sparse tensors will be "densified" before
+            computing numerical jacobians. The densified sparse tensor
+            is a copy of a sparse tensor with all unspecified elements
+            being materialized with zero value.
+        """
+        masked = masked_grad = kwargs.pop('masked', False)
+        sparse_layouts = {torch.sparse_coo, torch.sparse_csr, torch.sparse_csc, torch.sparse_bsr, torch.sparse_bsc}
+        STRIDED_REPRESENTATION = '__STRIDED_REPRESENTATION__'
+
+        def _convert_to_strided_representation(obj):
+            """Convert a differentiable non-strided tensor to a representation
+            containing differentiable strided tensors only.
+            """
+            if isinstance(obj, torch.Tensor) and obj.requires_grad:
+                d = dict(layout=obj.layout, shape=obj.shape, original=obj)
+                device = obj.device
+                if obj.layout is torch.sparse_coo:
+                    obj = obj.coalesce()
+                    indices, values = obj.indices(), obj.values()
+                    if masked:
+                        d.update(indices=indices)
+                        return (STRIDED_REPRESENTATION, d, values.requires_grad_(True))
+                    else:
+                        # Materialize unspecified elements with zero values
+                        full_nnz = obj.shape[:obj.sparse_dim()].numel()
+                        full_values = torch.zeros((full_nnz, *values.shape[1:]), dtype=values.dtype, device=values.device)
```

Comment:
Can't we use `obj.size()` to get the full size without this trick?
Or is the point to collapse all sparse dims similar to doing `.view(-1, *values.shape[1:])` ?

I'm not sure why you do this since it makes the indexing below hard.

Note that for both this one and the one below, can we use our add() function to get this in an easier way?

```python
full_val = torch.zeros(obj.size())
full_val = full_val + obj
```

Reply:
We can actually use other tricks to materialize COO tensor zeros:
```python
full_obj = torch.ones(obj.shape, dtype=obj.dtype, device=obj.device).to_sparse(layout=torch.sparse_coo, dense_dim=obj.dense_dim())
full_obj.values().sub_(1)
full_obj += obj
```
or
```python
full_nnz = obj.shape[:obj.sparse_dim()].numel()
full_indices = torch.ones(obj.shape[:obj.sparse_dim()], device=indices.device, dtype=torch.int8).nonzero().to(dtype=indices.dtype).T
full_values = torch.zeros((full_nnz, *obj.shape[obj.sparse_dim():]), device=values.device, dtype=values.dtype)
full_obj = torch.sparse_coo_tensor(full_indices, full_values, obj.shape)._coalesced_(True)
full_obj += obj
```
Performance-wise these appear to be equivalent, the former is a bit simpler, imho, so I'll pick this as a resolution.

Reply:
Unfortunately, the same approach does not work for CSR tensors because adding CSR tensors with batch dimensions is broken, see https://github.com/pytorch/pytorch/issues/107478.

Reply:
Applied the hints for COO tensors. For CSR tensors no changes are made because there exists bugs in adding CSR tensors.